components/engine_rocks_helper/src/sst_recovery.rs

```diff
@@ -0,0 +1,265 @@
+    damaged_files: Vec<FileInfo>,
+    max_hang_duration: Duration,
+    check_duration: Duration,
+}
+
+#[derive(Debug, Clone)]
+struct FileInfo {
+    // Corrupted file name. example: /000033.sst
+    name: String,
+    smallest_key: Vec<u8>,
+    largest_key: Vec<u8>,
+    // Time to generate recovery task, be used to record whether the timeout
+    start_time: Instant,
+}
+
+impl Runnable for RecoveryRunner {
+    type Task = String;
+
+    fn run(&mut self, sst: String) {
+        self.generate_scheduling_tasks(&sst);
+    }
+}
+
+impl RunnableWithTimer for RecoveryRunner {
+    fn get_interval(&self) -> Duration {
+        self.check_duration
+    }
+
+    fn on_timeout(&mut self) {
+        self.check_damaged_files();
+    }
+}
+
+impl RecoveryRunner {
+    pub fn new(
+        db: Arc<DB>,
+        store_meta: Arc<Mutex<StoreMeta>>,
+        max_hang_duration: Duration,
+        check_duration: Duration,
+    ) -> Self {
+        RecoveryRunner {
+            db,
+            store_meta,
+            damaged_files: vec![],
+            max_hang_duration,
+            check_duration,
+        }
+    }
+
+    fn generate_scheduling_tasks(&mut self, path: &str) {
+        if self.exist_scheduling_regions(path) {
+            return;
+        }
+
+        let live_files = self.db.get_live_files();
+        for i in 0..live_files.get_files_count() {
+            if path == live_files.get_name(i as i32) {
+                let f = FileInfo {
+                    name: live_files.get_name(i as i32),
+                    smallest_key: live_files.get_smallestkey(i as i32).to_owned(),
+                    largest_key: live_files.get_largestkey(i as i32).to_owned(),
+                    start_time: Instant::now(),
+                };
+
+                // only support recovering data key for now.
+                if !keys::validate_data_key(&f.smallest_key)
+                    || (!keys::validate_data_key(&f.largest_key)
+                        && f.largest_key.as_slice() != keys::DATA_MAX_KEY)
+                {
+                    self.set_panic_mark_and_panic(
+                        path,
+                        &format!(
+                            "key range mismatch, smallest key:{:?}, largest key:{:?}",
+                            &f.smallest_key, &f.largest_key
+                        ),
+                    );
+                }
+
+                // lock the store_meta and check if the range exist.
+                let _ = self.check_overlap_damaged_regions(f);
+
+                return;
+            }
+        }
+    }
+
+    // `sst_path` has been processed and is still in a damaged state.
+    fn exist_scheduling_regions(&self, sst_path: &str) -> bool {
+        self.damaged_files.iter().any(|f| f.name == sst_path)
+    }
+
+    // Periodically check for recorded damaged files。
+    //
+    // Acquire meta lock.
+    fn check_damaged_files(&mut self) {
+        let mut new_damaged_files = self.damaged_files.clone();
```

Comment:
Why clone?

Reply:
I guess you could make `check_overlap_damaged_regions` immutable, by pushing to `damaged_files` outside of it.

Reply:
yes if we make `check_overlap_damaged_regions` immutable, it still can't be called while `self` is mutable borrowed. of course we could pass struct fields instead of self into `check_overlap_damaged_regions`, but damaged_files are referenced internally to log

Reply:
Could you move the part of code that checks file against region ranges into a static function? (`fn get_overlapped_region(meta: &MutexGuard<..>, f: FileInfo) -> Vec<u64>`)

Reply:
Could you move the part of code that checks file against region ranges into a static method? (`get_overlapped_region(meta: &MutexGuard, f: FileInto) -> Vec<u64>`). You can then clone the meta to get around the immutable reference.

Reply:
At first I implemented exactly the same as you said, but there are still some places that needs to use `damaged_regions`, such as `set_panic_mark_and_panic`, which needs to print out all file information when panicking.
Also there shouldn't be any performance penalty for clone here as it is impossible that thousands of sst files exist at the same time.

Reply:
The code I mentioned (L140-L153) doesn't call `set_panic_mark_and_panic`. And yes, I agree the overhead is not a problem, I'm just not too comfortable with the code structure. Here I have to think hard to understand how some modifications inside `check_overlap_damaged_regions` is not effective (because the file list is cloned).

Reply:
> Here I have to think hard to understand how some modifications inside check_overlap_damaged_regions is not effective (because the file list is cloned).
Only modify when first time detection. Other times, you can think of the `damaged_files` as a snapshot?

> The code I mentioned (L140-L153) doesn't call set_panic_mark_and_panic. 
L122~L125 could access the file list early when timeout, if we retain `self.damaged_files` directly, maybe more flag vars are included and we may move L122~L125 to back.

Reply:
In [0458b9c](https://github.com/tikv/tikv/pull/10961/commits/0458b9c8397e2d3e9a518e77c49b6963a7323549) move out the modifications to `damged_files` to make the code clear.

Reply:
You are right, it's messier than I think. How about this:
```
let old_damaged_files = self.damaged_files.clone();
self.damaged_files.clear();
old_damaged_files.into_iter().for_each(|f| {
  if f.start_time... { panic }
  check_overlap_damaged_regions(f)
});
```

Reply:
If implement like this, files may be pushed again in `check_overlap_damaged_regions`. I think just make `check_overlap_damaged_regions` immutable and this also matches the naming of this function (only do check and don't modify `damaged_files` everywhere)

Reply:
Okay